### PR TITLE
Fix keras.ops.roll shift/axis handling to match numpy broadcasting

### PR DESCRIPTION
**Pull request by TheManishCode**
## Description

Fix `keras.ops.roll` shift/axis handling to match NumPy broadcasting behavior.

### Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/backend/openvino/numpy.py [hunk 1]
```diff
@@ -4142,7 +4142,15 @@ def reshape(x, newshape):
...
     if axis is not None:
-        result = ov_opset.roll(x, shift, axis).output(0)
+        # The Roll operation requires `shift` and `axis` to have the same
+        # length, while numpy broadcasts them against each other.
+        shifts = list(shift) if isinstance(shift, (list, tuple)) else [shift]
+        axes = list(axis) if isinstance(axis, (list, tuple)) else [axis]
+        if len(shifts) == 1:
+            shifts = shifts * len(axes)
+        if len(axes) == 1:
+            axes = axes * len(shifts)
```
[the remark this answers is not in the document]
Comment (TheManishCode) [reply keras/src/backend/openvino/numpy.py]: Good catch, thanks. Fixed in e7026ef: centralized the shift/axis broadcasting into a single `broadcast_shift_and_axis()` helper in `keras/src/backend/common/backend_utils.py`, which normalizes any array-like value (numpy arrays, etc.) via `.tolist()` before checking length, instead of the inline `isinstance(..., (list, tuple))` check that missed them. All four call sites (tensorflow/torch/openvino backends + `Roll.compute_output_spec`) now use this shared helper, and I added a test case with `np.array` shift/axis to `numpy_test.py` to cover it. This also surfaced and fixed a related crash: `Roll.compute_output_spec` was passing a numpy-array `axis` straight into `canonicalize_axes`, which blew up on `operator.index()`.

File: keras/src/backend/tensorflow/numpy.py [hunk 2]
```diff
@@ -2866,7 +2866,15 @@ def reshape(x, newshape):
...
     if axis is not None:
-        return tf.roll(x, shift=shift, axis=axis)
+        # `tf.roll` requires `shift` and `axis` to have the same length,
+        # while numpy broadcasts them against each other.
+        shifts = list(shift) if isinstance(shift, (list, tuple)) else [shift]
+        axes = list(axis) if isinstance(axis, (list, tuple)) else [axis]
+        if len(shifts) == 1:
+            shifts = shifts * len(axes)
+        if len(axes) == 1:
+            axes = axes * len(shifts)
```
[the remark this answers is not in the document]
Comment (TheManishCode) [reply keras/src/backend/tensorflow/numpy.py]: Good catch, thanks. Fixed in e7026ef: centralized the shift/axis broadcasting into a single `broadcast_shift_and_axis()` helper in `keras/src/backend/common/backend_utils.py`, which normalizes any array-like value (numpy arrays, etc.) via `.tolist()` before checking length, instead of the inline `isinstance(..., (list, tuple))` check that missed them. All four call sites (tensorflow/torch/openvino backends + `Roll.compute_output_spec`) now use this shared helper, and I added a test case with `np.array` shift/axis to `numpy_test.py` to cover it. This also surfaced and fixed a related crash: `Roll.compute_output_spec` was passing a numpy-array `axis` straight into `canonicalize_axes`, which blew up on `operator.index()`.

File: keras/src/backend/torch/numpy.py [hunk 3]
```diff
@@ -1823,7 +1823,17 @@ def reshape(x, newshape):
...
-    return torch.roll(x, shift, dims=axis)
+    if axis is not None:
+        # `torch.roll` requires `shifts` and `dims` to have the same length,
+        # while numpy broadcasts them against each other.
+        shifts = list(shift) if isinstance(shift, (list, tuple)) else [shift]
+        axes = list(axis) if isinstance(axis, (list, tuple)) else [axis]
+        if len(shifts) == 1:
+            shifts = shifts * len(axes)
+        if len(axes) == 1:
+            axes = axes * len(shifts)
```
[the remark this answers is not in the document]
Comment (TheManishCode) [reply keras/src/backend/torch/numpy.py]: Good catch, thanks. Fixed in e7026ef: centralized the shift/axis broadcasting into a single `broadcast_shift_and_axis()` helper in `keras/src/backend/common/backend_utils.py`, which normalizes any array-like value (numpy arrays, etc.) via `.tolist()` before checking length, instead of the inline `isinstance(..., (list, tuple))` check that missed them. All four call sites (tensorflow/torch/openvino backends + `Roll.compute_output_spec`) now use this shared helper, and I added a test case with `np.array` shift/axis to `numpy_test.py` to cover it. This also surfaced and fixed a related crash: `Roll.compute_output_spec` was passing a numpy-array `axis` straight into `canonicalize_axes`, which blew up on `operator.index()`.

File: keras/src/ops/numpy.py [hunk 4]
```diff
@@ -7361,6 +7361,17 @@ def call(self, x):
...
+            isinstance(self.shift, (list, tuple))
+            and isinstance(self.axis, (list, tuple))
+            and len(self.shift) != 1
+            and len(self.axis) != 1
+            and len(self.shift) != len(self.axis)
+        ):
+            raise ValueError(
+                "`shift` and `axis` must be broadcastable to the same "
+                f"length. Received: shift={self.shift}, axis={self.axis}"
+            )
```
[the remark this answers is not in the document]
Comment (TheManishCode) [reply keras/src/ops/numpy.py]: Good catch, thanks. Fixed in e7026ef: centralized the shift/axis broadcasting into a single `broadcast_shift_and_axis()` helper in `keras/src/backend/common/backend_utils.py`, which normalizes any array-like value (numpy arrays, etc.) via `.tolist()` before checking length, instead of the inline `isinstance(..., (list, tuple))` check that missed them. All four call sites (tensorflow/torch/openvino backends + `Roll.compute_output_spec`) now use this shared helper, and I added a test case with `np.array` shift/axis to `numpy_test.py` to cover it. This also surfaced and fixed a related crash: `Roll.compute_output_spec` was passing a numpy-array `axis` straight into `canonicalize_axes`, which blew up on `operator.index()`.

File: keras/src/ops/numpy_test.py [hunk 5]
```diff
@@ -6486,9 +6493,21 @@ def test_roll(self):
...
         self.assertAllClose(knp.roll(x, -1, axis=0), np.roll(x, -1, axis=0))
+        self.assertAllClose(
+            knp.roll(x, 1, axis=(0, 1)), np.roll(x, 1, axis=(0, 1))
+        )
+        self.assertAllClose(
+            knp.roll(x, (1, 2), axis=(0, 1)), np.roll(x, (1, 2), axis=(0, 1))
+        )
+        self.assertAllClose(
+            knp.roll(x, (1, 2), axis=(1,)), np.roll(x, (1, 2), axis=(1,))
+        )
```
[the remark this answers is not in the document]
Comment (TheManishCode) [reply keras/src/ops/numpy_test.py:6628]: Good catch, thanks. Fixed in e7026ef: centralized the shift/axis broadcasting into a single `broadcast_shift_and_axis()` helper in `keras/src/backend/common/backend_utils.py`, which normalizes any array-like value (numpy arrays, etc.) via `.tolist()` before checking length, instead of the inline `isinstance(..., (list, tuple))` check that missed them. All four call sites (tensorflow/torch/openvino backends + `Roll.compute_output_spec`) now use this shared helper, and I added a test case with `np.array` shift/axis to `numpy_test.py` to cover it. This also surfaced and fixed a related crash: `Roll.compute_output_spec` was passing a numpy-array `axis` straight into `canonicalize_axes`, which blew up on `operator.index()`.

File: keras/src/backend/common/backend_utils.py [hunk 6]
```diff
@@ -435,6 +435,31 @@ def to_tuple_or_list(value):
     return value
 
 
+def _roll_arg_to_list(value):
+    if isinstance(value, (list, tuple)):
+        return list(value)
+    if hasattr(value, "tolist"):
+        value = value.tolist()
+        return value if isinstance(value, list) else [value]
+    return [value]
```
Comment (hertschuh) [inline keras/src/backend/common/backend_utils.py]: Remove this and just reuse `to_tuple_or_list` instead.

File: keras/src/backend/common/backend_utils.py [hunk 7]
```diff
@@ -435,6 +435,31 @@ def to_tuple_or_list(value):
...
+
+def broadcast_shift_and_axis(shift, axis):
+    """Broadcast `shift` and `axis` arguments of `roll` to the same length.
+
+    Mirrors numpy's behavior of broadcasting a scalar (or length-1
+    sequence) `shift`/`axis` against the other argument. Accepts ints,
+    lists, tuples, and array-likes (e.g. numpy arrays) exposing `tolist()`.
+    """
+    shifts = _roll_arg_to_list(shift)
+    axes = _roll_arg_to_list(axis)
```
Comment (hertschuh) [inline keras/src/backend/common/backend_utils.py]: Use `to_tuple_or_list` instead.

File: keras/src/backend/common/backend_utils.py [hunk 8]
```diff
@@ -435,6 +435,31 @@ def to_tuple_or_list(value):
...
+    if len(shifts) == 1:
+        shifts = shifts * len(axes)
+    if len(axes) == 1:
+        axes = axes * len(shifts)
+    return shifts, axes
+
```
Comment (hertschuh) [inline keras/src/backend/common/backend_utils.py:459]: Shouldn't you raise here if `shifts` and `axes` don't have the same length?

File: keras/src/ops/numpy.py [hunk 9]
```diff
@@ -7360,7 +7361,14 @@ def call(self, x):
...
         if self.axis is not None:
-            canonicalize_axes(self.axis, len(x.shape))
+            shifts, axes = broadcast_shift_and_axis(self.shift, self.axis)
+            canonicalize_axes(axes, len(x.shape))
+            if len(shifts) != len(axes):
+                raise ValueError(
+                    "`shift` and `axis` must be broadcastable to the same "
+                    f"length. Received: shift={self.shift}, "
+                    f"axis={self.axis}"
+                )
```
Comment (hertschuh) [inline keras/src/ops/numpy.py]: This can be moved to `broadcast_shift_and_axis` to also apply in the non-functional case.

File: keras/src/backend/common/backend_utils.py [hunk 10]
```diff
@@ -435,6 +435,31 @@ def to_tuple_or_list(value):
...
+def _roll_arg_to_list(value):
+    if isinstance(value, (list, tuple)):
+        return list(value)
+    if hasattr(value, "tolist"):
+        value = value.tolist()
+        return value if isinstance(value, list) else [value]
+    return [value]
+
+
+def broadcast_shift_and_axis(shift, axis):
```
Comment (hertschuh) [inline keras/src/backend/common/backend_utils.py]: Rename `normalize_shift_and_axis`, to me "broadcast" is something a bit different.

**Review by hertschuh — CHANGES_REQUESTED** [review]
Thanks for looking into this!

**Comment by TheManishCode** [comment]
Re: the failing **Zizmor** check ([run 29908230655](https://github.com/keras-team/keras/actions/runs/29908230655)) — this is **not caused by this PR**.

- This PR only touches `keras/src/**` (the `roll` op and its backends/tests); it changes no `.github/workflows/*.yml` files.
- Every Zizmor finding is a warning of the same kind — *"action's hash pin has mismatched or missing version comment: points to commit `3d3c42e5aac5`"* — across `actions.yml`, `nightly.yml`, `scorecard.yml`, `gpu_tests.yml`, `labeler.yaml`, `tpu_tests.yml`, `zizmor.yml`, etc. These are workflow-infrastructure issues, not code issues.
- The same Zizmor check is currently failing on **`master` push builds** and on **all other open PRs**, so it's a pre-existing repo-wide failure rather than anything introduced here.

The checks relevant to this change pass locally: the `roll` unit tests pass on both TensorFlow and PyTorch backends (including the new numpy-array `shift`/`axis` cases), and `ruff check`/`ruff format` are clean.

I've left the workflow files untouched, since pinning those actions is a maintainer-side infra fix and out of scope for this bugfix. Happy to rebase once the pin mismatch is resolved on `master`.

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!

**Comment by TheManishCode** [comment]
Thank you! I really appreciate the guidance and the review. I hope to keep learning through open source and contribute more often. Looking forward to contributing again!
